### ISSUE TITLE
Generate JUnit sources from .feature files

### ISSUE THREAD

**Issue by aslakhellesoy** [cucumber/cucumber-jvm#1487]
Let's discuss whether we should use a code-generation approach for JUnit going forward.

Potential benefits:
* Users can use Cucumber with *any* test runner
* No need to maintain special runners for JUnit/TestNG/Spock etc

Potential drawbacks:
* IDE plugins would have to be updated to invoke the generator before compiling generated classes
* We might have to write a Maven/Gradle plugins for code generation

We'd bundle templates for popular test runners, providing a way to use their own templates, or tweak built-in ones.

SpecFlow uses a code generation approach. (SpecFlow generates NUnit / xUnit files from `.feature`files). See `HearShoutFeature.cs` in [this project](https://github.com/cucumber-ltd/shouty.net/tree/master/Shouty.Specs) to get an idea.

To make this work we'd have to write an API that generated Java files can use. If we were to explore this strategy I would do it in the following order:

1. Develop an API
2. Hand-write a few JUnit4 and JUnit5 classes that use this API, to get a feel for what the code generator needs to do
3. Tweak the API based on what we learn in 2.
4. Implement a code generator.

This approach could live side-by-side with the existing runner - we don't have to commit to this approach if it turns out to be impractical.

**Comment by aslakhellesoy** [comment]
@gasparnagy @SabotageAndi I'd love to get some advice from you hare. What pros/cons have you experienced in SpecFlow with the code generation approach? Do you think this could work well for Java too?

**Comment by mpkorstanje** [comment]
In general I can see the benefit of going through Cucumbers architecture and looking at way to reduce the cost of implementing adapters. However I think that this particular solution puts too many constraints on other aspects. Or if implemented next to the current runners, provides no advantages at the cost of additional complexity.

The amount of complexity added to IDE plugins is non trivial. More so then what you've mentioned. Currently selecting a scenario is as simple as selecting a (line in a) feature file and providing that to Cucumber. Using generated code an IDE would now have to determine the settings of the code generator from `pom.xml` or `build.gradle`, generate the code, match is back to the feature file, and select that test for execution. Relying on implementation details all the way through. 
This is not a situation we should find ourselves in. IDE's should have a clear interface to discover, select, execute and report on the execution of Cucumber scenarios. Currently the CLI and plugin API provides some of this services but we can do much better here.

Likewise users of Cucumber also need a clear way to select which scenarios are executed. We currently provide line, name and tag filters for this. If we go the code generation route this is no longer possible. We can only express this in terms of the test framework in which the code is generated and this may not be a good mapping.

I also think the actual number of test runners is limited. We currently have Junit4 and TestNG. Spock uses JUnit as runner. Expressing Cucumber tests in Spock should result in the same outcome as expressing them in JUnit. 
So this  seems like a high effort for a low value of n. Especially since most of the issues with JUnit and TestNG came from using the wrong abstractions internally -- we used `Runtime` rather then its components leading to all sorts of conceptual mismatches. 

As for JUnit 5, it is not possible to implement custom test runners in JUnit 5 in the way we do for JUnit 4 and TestNG. Rather it provides a `TestEngine` framework. This test engine framework provides IDE's and other tooling with a clear interface to discover, select, execute and report on the execution of Cucumber scenarios. Which is exactly what we need at the moment.

Nevertheless, something is motivating you. What makes you experience the current integration as high effort? Or clunky.

**Comment by gasparnagy** [comment]
For SpecFlow, the choice of code generation was not an option, but a must, because at that time the popular .NET unit test runners did not provide an interface for dynamic test generation (like the one that Cucumber uses with JUnit). Even today this is only available for xUnit, where we allow users to avoid the code generation (optionally).

There are a few good points in the generator approach but also a lot of obstacles. Over the years it turned out that doing the generation design-time (i.e. when you save the file) or anyhow depending on the IDE integration is a continuous pain and complexity. Currently, we try to move away from this and generate the source files (called code-behind files in SpecFlow) during build (in our case MsBuild). (For SpecFlow v3 we have even considered to drop it completely, but finally we kept it for backwards compatibility and to support wider range of test runners.)

I have written a post about this once: http://gasparnagy.com/2017/04/specflow-without-code-behind-files/

Here is my pro/con summary, if that helps:

Pro:

- Easy integration with any test runner / build system that supports the unit test framework you are generating for.
- No need to add any additional classes to the project by the users (like the annotated JUnit class)
- Easier to support different programming languages within the platform (e.g. SpecFlow supports C# and VB.NET)
- Easier to make customizations (e.g. special, domain/project specific annotations on the generated unit test class/method) via generator plugins
- Provides an easy way to debug feature files (allow setting break points in feature files) -- at least in .NET you can solve this easily by generating special lines to the source code

Con:

- Implementing/maintaining the generator code is not trivial, especially because the programming languages keep evolving
- Integrate the generation to the IDE/build pipeline is a continouos effort
- People might need to commit the generated code (if generated IDE time) and suffer from tricky merge issues
- Generation needs to be configurable (e.g. which unit test fw/language to generate), so you need to find an option to specify Cucumber configuration that is also readable by the generator. 
- Generating the necessary class for Scenario Outlines might be bit tricky (depending on the capabilities of the unit test runner you generate for).

**Comment by aslakhellesoy** [comment]
Thanks for sharing your thoughts @mpkorstanje and @gasparnagy - very useful.

The reasons I'm exploring this as a possibility are:

a) People keep asking about using various JUnit annotations that aren't compatible with our current runner.
b) People keep asking about a JUnit 5 runner, and it seems to me the main obstacle is for someone to find the time to learn the internal JUnit 5 API (as well as Cucumber API) and then sit down and write it.

In summary, I wanted to figure out if we can reduce some of this complexity. I suggest we leave this open until stalebot closes it. I understand the drawbacks, but I might get inspired to do a spike so I can try it out. Or I might not :-)

**Comment by mpkorstanje** [comment]
> a) People keep asking about using various JUnit annotations that aren't compatible with our current runner.

It is a rather confusing situation. When using the runner we have two different test contexts. The runner class is JUnits test context. The glue classes are Cucumbers test context. All class rules work in the JUnit context. Method rules do not because the runner class is never instantiated. The glue classes are however. But they are in the wrong context.

This would have to be solved in the object factory. Unfortunately most of JUnits rules don't work well with multiple classes.

To the casual observer this difference is not clear or obvious though. Walks like a duck...

Code generation would solve the problem of not having a method to apply the rule too but the rule would not be accessible from inside the glue. So of limited use.

I think this is a good reason to move away from runners written in Java code. Ideally people just plugin a maven or gradle dependency/plugin and have their tests ran.

> b) People keep asking about a JUnit 5 runner, and it seems to me the main obstacle is for someone to find the time to learn the internal JUnit 5 API (as well as Cucumber API) and then sit down and write it. 

I learned allot about test engine design from my initial JUnit 5 spike. JUnit 5 has been build with all the lessons from JUnit 4 in mind. The result is worth studying, most lessons apply to Cucumber too.

Give the TestEngine API and JUnit implementation of it a read! Migth be inspiring!
